### DISCUSSION TITLE
The Overloading Passage

### DISCUSSION THREAD

**Discussion by ghost** [h3rald/min#99] [Features: On Hold]
When i trying the new feature, generics, then i realized that generics are a way to handle with **overloading problem**:
```
( symbol len
	((num|string :t) :x  t :y)
	(
		x type :str_t
		
		(str_t "num" ==)
		(...)
		when
		
		(str_t "string" ==)
		(...)
		when
	)
) ::

6 8                  len puts ; 10
"LAN" "TERN" len puts ; 7
```
But that wasn't enough yet. Because **cross type implementing** cannot be done within generics:
```
len:
string int => posint ;positive integer
int string => posint

; Can you write a generic for above parameter patterns?
```
I thought a more general idea that we can declare them as quots:
```
( symbol len
	(
		(string :str)     :id_1
		(num :x num :y) :id_2
		(dict :dc)         :id_3
	)
	(
		(id_1 true ==)
		(...)
		when
		
		(id_2 true ==)
		(...)
		when
		
		(id_3 true ==)
		(...)
		when
	)
) ::

"lololololo" len puts ; 10
3 4           len puts ; 5
{1 :a  2 :b} len puts ; 2
```
We can use above `id`s for understanding what pattern will we used. `id`s are boolean typed values.

So, this is a passage for expanding *genericism* to *parameter-patterns* system.

Good idea??

**Comment by h3rald** [comment]
I understand what you are trying to do, and yes, it would be nice to have a way to overload symbols, i.e. make them support different signatures.

Technically it could be done, and it would be implemented as something like this:

1. If there is a signature to try, try a signature:
   - if no error is thrown, go to 3.
   - if an error is thrown, restore the stack and go back to 1.
2. If no signature is compatible, throw an error.
3. If a signature is compatible, execute the operator body.

There are however some things I don't like very much:
- The fact that I have to try different signatures by catching errors. But again, it's not too different from what already happens with expectation, the only difference is that in this case there are "multiple tries" before failing.
- The syntax is not great... but maybe we can work on that! I think you just forgot some `==>`...


Maybe something like this:

```scheme
(
  symbol overload-test
  (
   (int :i str :s ==> str :result)
   (str :s int :i ==> str :result)
  )
  (
    (signature 0 ==)
      (...)
    when
    (signature 1 ==)
      (...)
    when
  )
) ::
```

Instead of binding to boolean values, we could use a special symbol `signature` that is set to the index of the matching signature.

What do you think?

The way to detect multiple signatures would simply be detecting multiple quotations (and only quotations) within the signature quotation, each with a `==>` symbol in it.

**Reply by ghost** [reply]
@h3rald 
> The syntax is not great... but maybe we can work on that! I think you just forgot some ==>...

Yep, i forgot to write `==>`. While i was writing, i have focused to idea, so i forgot it.

> Technically it could be done, and it would be implemented as something like this:

The algorithm is what we thought about it. We're on this algorithm.

> But again, it's not too different from what already happens with expectation

Do you mean `expect`? No, this is difference because **expectations** not working with **typeclasses**. So a hacky-doing i did again 😄.

> What do you think?

My idea has been reshaped by you and your shaping is enough to make that able. If another problem appears about the main idea then ask me.

**Comment by ghost** [comment]
What role generics should have in this idea??

**Reply by h3rald** [reply]
> Do you mean expect? No, this is difference because expectations not working with typeclasses. So a hacky-doing i did again 😄.

...oh but `expect` works with type classes... now 😉 

> What role generics should have in this idea??

Well, I have to be careful to make sure I am parsing the signature quotation properly. I have to handle a special case that if it is composed by only quotations then it is basically a list of multiple signatures. At present I check if there's a quotation in some places (odd position for input, even position for output) then I verify that it contains a valid generic definition... I must make sure they don't get mixed up! (Also a signature alternative can contain generics...)

It has nothing to do with overloading, just with parsing the signature quotation.

**Reply by ghost** [reply]
@h3rald 
> ...oh but expect works with type classes... now 😉

Yeah. I controlled in new v0.29.1 Min version and expectations worked !! Really, thanks for. 👍 

> It has nothing to do with overloading, just with parsing the signature quotation.

Parsing the signature quotation *makes* overloading with symbols. This is a brilliant idea. It is a way ended at **overloading**.

> Also a signature alternative can contain generics...

However this can make some conflicts:
```
...
((string|num :t) :x  t :y ==> int :res) ; --> 0
(string :z  string :w ==> int :res) ; --> 1
...
```

Which one must be chosen for `string string` pattern?

Same thing below:

```
...
(string|num :x  string|num :y ==> int :res) ; --> 0
(string :z  string :w ==> int :res) ; --> 1
...
```

Programmer may be informed about possible died signature quotations or possible ambiguous signature quotations.

**Reply by h3rald** [reply]
Well, I was thinking that simply the first matching signature is used... honestly I don't think you'd have similar signatures really! They should ideally be more or less mutually exclusive, or anyway different.

**Reply by ghost** [reply]
> Well, I was thinking that simply the first matching signature is used... honestly I don't think you'd have similar signatures really! They should ideally be more or less mutually exclusive, or anyway different.

Maybe the programmer wrote similar signature quots mistakenly. He should be warned about this. After doing that, he will be able to fixing his mistakes.

Am i right?

**Reply by h3rald** [reply]
Yes... you are right. The only problem is to figure out how to do so, I'd have to have to check if a signature "matches" another one already. In your example I think I can do it, but if you use custom type classes I can't just by name, I'd have to re-test every value again and I wouldn't do that... but it would definitely be possible with things like unions vs simple types, or int and num etc.

**Reply by ghost** [reply]
There are so many ideas at the discussion. You should make a combine of them with yours ofcourse 👍 to see what must we change on our ideas.

**Reply by h3rald** [reply]
Oh I will... 

On another note, if I may ask... I am curious: where are you from? I understand your way of talking but he construction of the sentences is not 100% English.

**Reply by ghost** [reply]
😄 😄 😄 
Hmm.m.m.m.mm... What languages did you see in my sentences????? ❓ Not %100 English?? 😃  How construction i have in my talking, in my texting  etc. ... ?

I don't understand sometimes what i talked, texted. Because English has a structured syntax. I use it. You may think who i am an English programmer 😆 . Maybe my English vocabulary lacks... Maybe i have a cluttered vocabulary... Nevermind. Just understand what i mean of. 👍  (Or don't understand. It doesn't matter 🤣 )

**Reply by h3rald** [reply]
Oh no your English vocabulary is *excellent* (better than the average native English speaker 🤣), just frequent use of gerunds for example is uncommon in native English, and other constructions I can't quite pinpoint...

Let's see... not a Latin-derived language otherwise I would have noticed 😉

I am going to take a wild guess and try... Turkish??

OK, I admit I googled your nickname... 🦉 right?? 😊

**Reply by ghost** [reply]
Okay, okay. You won the battle!

> Oh no your English vocabulary is excellent 

A grandpa says: "Don't lie me son"

> Turkish??

Yes, I am Turkish.

> Better than the average native English speaker 🤣

No, I am not better than the average native English speaker. You cannot infer my speaking style from my texts, vice versa isn't possible too.

> right?? 

Yes you're right on nearly all your sayings, searchings, you are so great. Go on!

**Reply by h3rald** [reply]
Are you gone? Just noticed that you seem to have deleted your account on GitHub. If I offended you anyway, I sincerely apologize, it wasn't my intention.

it is very unusual for me to meet someone so interested about one of my open source projects, thank you for your contributions to this micro-community. I believe your ideas helped improving min, and I am truly sad to see you disappear so suddenly.

I really hope that you will be back in some form, as I really enjoyed our discussions.
